src/OpenSilver.ControlsKit.Controls/AdaptiveColumnsPanel.cs

```diff
@@ -0,0 +1,198 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Windows;
+using System.Windows.Controls;
+
+namespace OpenSilver.ControlsKit
+{
+    /// <summary>
+    /// A responsive <see cref="Panel"/> that dynamically switches between layouts based on its width:
+    /// when its width meets or exceeds <c>NoColumnsBelowWidth</c>, children are arranged into equal-width columns (one per visible child);
+    /// otherwise, they are stacked vertically in a single column.
+    /// Child horizontal and vertical alignments are respected within each allocated slot.
+    /// </summary>
+    public class AdaptiveColumnsPanel : Panel
+    {
+        /// <summary>
+        /// When the available width is ≤ this threshold, children stack vertically.
+        /// When > this threshold, children lay out in N equal-width columns (N = # of children).
+        /// </summary>
+        public static readonly DependencyProperty NoColumnsBelowWidthProperty =
+            DependencyProperty.Register(
+                nameof(NoColumnsBelowWidth),
+                typeof(double),
+                typeof(AdaptiveColumnsPanel),
+                new FrameworkPropertyMetadata(
+                    500d,
+                    FrameworkPropertyMetadataOptions.AffectsMeasure));
+
+        /// <summary>
+        /// When the available width is ≤ this threshold, children stack vertically.
+        /// When > this threshold, children lay out in N equal-width columns (N = # of children).
+        /// </summary>
+        public double NoColumnsBelowWidth
+        {
+            get => (double)GetValue(NoColumnsBelowWidthProperty);
+            set => SetValue(NoColumnsBelowWidthProperty, value);
+        }
+
+        // Get visible children only once and as FrameworkElement directly
+        private List<FrameworkElement> GetVisibleChildren() =>
+            Children.OfType<FrameworkElement>()
+                    .Where(c => c.Visibility != Visibility.Collapsed)
+                    .ToList();
+
+        // Determine layout mode in one place
+        private bool ShouldUseColumns(double availableWidth, int childCount) =>
+            !double.IsInfinity(availableWidth) &&
+            availableWidth > NoColumnsBelowWidth &&
+            childCount > 0;
+
+        protected override Size MeasureOverride(Size availableSize)
+        {
+            double layoutWidth = double.IsNaN(this.Width) ? availableSize.Width : this.Width;
+            var children = GetVisibleChildren();
+            int count = children.Count;
```

Comment:
Avoid divide by zero exception line 77
```suggestion
            int count = children.Count;
            if (count == 0)
            {
                return base.MeasureOverride(availableSize);
            }
```